doc/designs/fips_indicator.md

```diff
@@ -0,0 +1,266 @@
+
+Requirements
+------------
+
+The following information was extracted from the FIPS 140-3 IG [5] “2.4.C Approved Security Service Indicator”
+
+- A module must have an approved mode of operation that requires at least one service to use an approved security function (defined by [2] and [3]).
+- A FIPS 140-3 compliant module requires a built-in service indicator capable of indicating the use of approved security services
+- If a module only supports approved services in an approved manner an implicit indicator can be used (e.g. successful completion of a service is an indicator).
+- An approved algorithm is not considered to be an approved implementation if it does not have a CAVP certificate or does not include its required self-tests. (i.e. My interpretation of this is that if the CAVP certificate lists an algorithm with only a subset of key sizes, digests, and/or ciphers compared to the implementation, the differences ARE NOT APPROVED. In many places we have no restrictions on the digest or cipher selected).
+- Documentation is required to demonstrate how to use indicators for each approved cryptographic algorithm.
+- Testing is required to execute all services and verify that the indicator provides an unambiguous indication of whether the service utilizes an approved cryptographic algorithm, security function or process in an approved manner or not.
+- The Security Policy may require updates related to indicators. AWS/google have added a table in their security policy called ‘Non-Approved Algorithms not allowed in the approved mode of operation’. An example is RSA with a keysize of < 2048 bits (which has been enforced by [4]).
+
+Since any new FIPS restrictions added could possibly break existing applications
+the following additional OpenSSL requirements are also needed:
+
+- The FIPS restrictions should be able to be disabled using Configuration file options (This results in unapproved mode and requires an indicator).
+- A mechanism for logging the details of any unapproved mode operations that have been triggered (e.g. DSA Signing)
+- The FIPS restrictions should be able to be enabled/disabled per algorithm context.
+- If the per algorithm context value is not set, then the  Configuration file option is used.
+
+Solution
+--------
+
+In OpenSSL most of the existing code in the FIPS provider is using
+implicit indicators i.e. An error occurs if existing FIPS rules are violated.
+
+The following rules will apply to any code that currently is not FIPS approved,
+but needs to be.
+
+- The fipsinstall application will have a configurable item added for each algorithm that requires a change.
+These options will be passed to the FIPS provider in a manner similar to existing code.
+
+- A user defined callback similar to OSSL_SELF_TEST will be added. This callback
+will be triggered whenever an approved mode test fails.
+It may be set up by the user using
```

Comment:
If a user defined callback is used, then there should be also a pointer for user defined context, otherwise user application will have no way to determine what is calling the callback in any complex program that uses many libraries and threads or async programming paradigms, here the application flow is not linear.

Reply:
OSSL_INDICATOR_set_callback(libctx, cb, arg) allows you to change the cb and additional arg...
OSSL_INDICATOR_set_callback(libctx, NULL, arg) can be used to change the arg also.

Reply:
I do this in evp_test.c in the example dsa PR.

Reply:
I guess what I do not understand is why use a callback.

Even for testing why not simply have a param on the ctx that reports on the status and can simply be queried with a standard get_params after the operation is complete ?

From the FIPS pov it does not matter what error happened, the indicator just need to indicate if the service (as a whole) was approved or not approved. Receiving multiple notifications seem technically out of the standard, although probably not prohibited, yet kinda wasteful.

For the vast majority of applications what matters is to know if the service they just requested was approved or not, all the other details are not needed.

It is much easier and less prone to error to "consult" a context than to deal with callbacks for application writers.
Especially for cases where you need to deal with language bindings, where dealing with callbacks is really hard due to the FFI layer in between.

Reply:
The callback has two purposes:

1. notifying callers as to exactly what non-approved operations their application is making and
2. providing a way for non-approved operations to be allowed to continue.

Agencies that have to use FIPS but also have non-compliant usage need to be able to audit they use -- hence the logging aspect to the callback.  They also need an exception mechanism to allow such through.

The fundamental premise here is that that applications are _not_ going to check indicators and that operations should, by default, fail if non-approved.  Q: how many applications have added post operation FIPS indicator checking throughout RedHat's sources?  I'd hazard a guess that it's fairly close to zero.  By producing an error, we're forcing compliance.  Being non-compliant stops being an accident and becomes a deliberate decision.  We need to support such a decision both via config and programmatically.

Reply:
Producing an error is just fine, and having a toggle to allow relaxed operation is also fine, none of this predicates a callback system.

Applications by themselves are not going to nitpick what part of the service failed (nor FIPS requires that). A log is useful at most during the debugging, once an application is in production either you allow to use a service that claims unapproved or you don't.
Therefore to achieve what applications needs, strictly speaking you need only two things:
1. a parameter to pass at context creation that toggles from hard fail to allow the operation thorugh even if not approved.
2. a parameter to consult on the context to optionally check if the operation was actually approved or not approved

The rest here is "cool" but unnecessary and a lot harder to deal with for applications, which now have to set up a whole callback system when they need to allow unapproved services that will absolutely not look at the content returned by this callback and will just allow anything through in basically all cases.

Note the above is necessarily true for two reasons: applications try to be compatible with different versions of OpenSSL, certifications and different providers, so they can't count on seeing any specific error, the common denominator is just a binary: approved/not-approved for the whole service. (I intend to allow FIPS certified pkcs#11 modules to be used in FIPS mode via the pkcs11-provider and report indicators too, but will not be able to report the same errors the OpenSSL upstream FIPS module will anyway). 

Note that I am not against the use of callbacks if it is an optional, for specific case to gain additional information, as long as (1) and (2) can be achieved w/o them, what I think is excessive is that the callback system seem the only way to achieve this goal, as it is a very heavy way to go about it for normal applications.

Reply:
The use case you are missing is an application that calls into a third party library that calls OpenSSL.  How does the application convince the third party library to pass an additional parameter to OpenSSL indicating that non-approved operation is okay?  It can't without the callback.

Reply:
I am not sure that is a problem that callbacks will solve properly, how do you know which call called the callback?

If you do not tie the callbacks to a libctx than you have no way to know what are you permitting because your callback can be called by anything. And/or you may actually interfere with libraries trying to set *their* callback because they know they are using an operation in a non-fips relevant setting. You can't set two callbacks after all, and depending on the sequence of calls to the library you have no good way to predict which one will end up being the last set.

So you *have to* tie the callbacks to a specific libctx and avoid interference between different libraries and application, and if you do tie the callback to a libctx then you will still not be able to catch a library that (correctly so) initialized their own libctx.

The most likely solution for those cases are:
- the application/admin just sets a global relaxed configuration in openssl.cnf and just accept it is using unapproved algorithms throughout.
- the developer/distribution will appropriately patch the offending library to either expose a knob or not use the unapproved algorithm, or set the local context relaxation parameter in the library.

As I said, I am not against having callbacks for special cases, the problem is if that's the only way. It is cumbersome and your example actually show it may violate the separation between application and libraries or separate libraries if done improperly, with unpredictable outcomes (and very hard to debug ones at that).

Reply:
This seems a very long discussion. The callbacks are a logging mechanism.

If you want to query after every single operation to check FIPS approved status you can (although I cant imagine many users will actually be doing this is practise).